### PR TITLE
[reflection] Push MonoError through misc methods

### PR DESCRIPTION
**Pull request by lambdageek**

### CODE REVIEW

File: mono/metadata/reflection.c [hunk 1]
```diff
@@ -1164,6 +1164,8 @@ method_encode_code (MonoDynamicImage *assembly, ReflectionMethodBuilder *mb)
 	guint32 header_size = 12;
 	MonoArray *code;
 
+	mono_error_init (error);
+
 	if ((mb->attrs & (METHOD_ATTRIBUTE_PINVOKE_IMPL | METHOD_ATTRIBUTE_ABSTRACT)) ||
 			(mb->iattrs & (METHOD_IMPL_ATTRIBUTE_INTERNAL_CALL | METHOD_IMPL_ATTRIBUTE_RUNTIME)))
 		return 0;
```
Comment (kumpera) [inline mono/metadata/reflection.c]: Zero is valid return value.

We should document that the caller must not check the return value for errors.

File: mono/metadata/reflection.c [hunk 2]
```diff
@@ -1182,10 +1184,10 @@ method_encode_code (MonoDynamicImage *assembly, ReflectionMethodBuilder *mb)
 		if (code == NULL){
 			char *name = mono_string_to_utf8 (mb->name);
 			char *str = g_strdup_printf ("Method %s does not have any IL associated", name);
-			MonoException *exception = mono_get_exception_argument (NULL, "a method does not have any IL associated");
+			mono_error_set_argument (error, NULL, "a method does not have any IL associated");
 			g_free (str);
 			g_free (name);
-			mono_raise_exception (exception);
+			return 0;
```
Comment (kumpera) [inline mono/metadata/reflection.c]: Could we return negative for error as a way to address the previous issue?

Comment (lambdageek) [inline mono/metadata/reflection.c]: It's a `guint32` return value.  I have no idea if there can really be that many codes (I don't see why not).

File: mono/metadata/reflection.c [hunk 3]
```diff
@@ -3808,97 +3900,125 @@ compare_types_by_table_idx (MonoReflectionTypeBuilder **type1, MonoReflectionTyp
...
 	int i;
+
+	mono_error_init (error);
 	
-	mono_image_add_cattrs (assembly, tb->table_idx, MONO_CUSTOM_ATTR_TYPEDEF, tb->cattrs);
+	if (!(succeeded = mono_image_add_cattrs (assembly, tb->table_idx, MONO_CUSTOM_ATTR_TYPEDEF, tb->cattrs, error)))
+		goto leave;
 	if (tb->fields) {
-		for (i = 0; i < tb->num_fields; ++i) {
+		for (i = 0; succeeded && i < tb->num_fields; ++i) {
```
Comment (kumpera) [inline mono/metadata/reflection.c]: Why can't we fail fast here? This code looks complicated.

**Comment by lambdageek** [comment]
@monojenkins merge

**Pull request merged**